File: lib/data/repositories/ai_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class AIService {
  static const String apiKey = "AIzaSyApqhdLF8wa3gKufi56lcljgH5pmTs1PvY";
  static const String apiUrl =
      "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=$apiKey";

  Future<List<String>> generateWords() async {
    try {
      final response = await http.post(
        Uri.parse(apiUrl),
        headers: {"Content-Type": "application/json"},
        body: jsonEncode({
          "contents": [
            {
              "parts": [
                {"text": "Generate 6 random common words."}
              ]
            }
          ]
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data.containsKey("candidates")) {
          final text = data["candidates"][0]["content"]["parts"][0]["text"];
          return text.split(" "); // Convert the response into a list of words
        } else {
          throw Exception("Invalid response format: $data");
        }
      } else {
        throw Exception("API Error: ${response.statusCode} - ${response.body}");
      }
    } catch (e) {
      print("Gemini API Error: $e");
      throw Exception("Failed to fetch words");
    }
  }
}

File: Downloads/BackendPIM-integration_khlaf/src/news/ai.service.ts
import axios from 'axios';
import { Injectable } from '@nestjs/common';

@Injectable()
export class AIService {
  private readonly apiKey = 'YOUR_GEMINI_API_KEY';
  private readonly geminiURL = `https://generativelanguage.googleapis.com/v1/models/gemini-pro:generateText?key=${this.apiKey}`;

  async summarizeNews(articles: string[]) {
    const API_KEY = process.env.GEMINI_API_KEY;
    const url = `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=${API_KEY}`;
  
    const prompt = `Summarize these news articles: ${JSON.stringify(articles)}. 
    Return JSON format: [{"title": "News Title", "summary": "Short summary"}]`;
  
    try {
      const response = await axios.post(url, {
        contents: [{ parts: [{ text: prompt }] }],
      });
  
      const summaries = JSON.parse(response.data.candidates?.[0]?.content?.parts?.[0]?.text.trim());
      return summaries;
    } catch (error) {
      console.error("Gemini API Error:", error.response?.data || error.message);
      return { error: "Failed to summarize news." };
    }
  }

  async classifySentiment(articleText: string): Promise<string> {
    const prompt = `Classify this MS news article as positive, neutral, or negative: ${articleText}`;
    const response = await axios.post(this.geminiURL, { prompt });
    return response.data?.candidates?.[0]?.text ?? 'neutral';
  }

  async rankNews(newsList: any[], userInterest: string) {
    const API_KEY = process.env.GEMINI_API_KEY;
    const url = `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=${API_KEY}`;
  
    const prompt = `Rank these MS news articles from most relevant to least relevant based on the user's interest: ${userInterest}. 
    Articles: ${JSON.stringify(newsList)}.
  
    Return ONLY valid JSON in the exact format:
    [{"title": "News Title", "rank": 1}, {"title": "Another News Title", "rank": 2}]
    
    Do NOT include any Markdown formatting, code blocks, explanations, or extra text.`;
  
    try {
      const response = await axios.post(url, {
        contents: [{ parts: [{ text: prompt }] }],
      });
  
      let aiResponse = response.data.candidates?.[0]?.content?.parts?.[0]?.text.trim();
  
      // 🚨 Remove unwanted Markdown code block markers (` ```json ... ``` `)
      if (aiResponse.startsWith("```json")) {
        aiResponse = aiResponse.replace(/```json|```/g, "").trim();
      }
  
      // ✅ Parse clean JSON response
      const rankedNews = JSON.parse(aiResponse);
      return { ranked_news: rankedNews };
    } catch (error) {
      console.error("Error parsing AI response:", error);
      return { error: "Invalid AI response format." };
    }
  }
}
